test/e2e/gateway_api_test.go

```diff
@@ -423,6 +428,286 @@ func testGatewayAPIRBAC(t *testing.T) {
+
+	if err := assertExpectedDNSRecords(t, map[expectedDnsRecord]bool{
+		{dnsName: "baz." + domain + ".", gatewayName: "test-gateway-update"}: true,
+		{dnsName: "foo." + domain + ".", gatewayName: "test-gateway-update"}: false,
+		{dnsName: "bar." + domain + ".", gatewayName: "test-gateway-update"}: false,
+	}); err != nil {
+		t.Fatalf("DNSRecord expectations not met: %v", err)
+	}
+
+	// Delete the gateway
+	if err := kclient.Delete(context.TODO(), gateway); err != nil {
+		if errors.IsNotFound(err) {
+			return
+		}
+		t.Errorf("failed to delete gateway %q: %v", gateway.Name, err)
+	}
+
+	// Check the DNSRecord gets deleted for the remaining listener hostname
+	if err := assertExpectedDNSRecords(t, map[expectedDnsRecord]bool{
+		{dnsName: "baz." + domain + ".", gatewayName: "test-gateway-update"}: false,
+	}); err != nil {
+		t.Fatalf("DNSRecord expectations not met: %v", err)
+	}
+}
+
+func testGatewayAPIDNSListenerWithNoHostname(t *testing.T) {
+	gatewayClass, err := createGatewayClass(operatorcontroller.OpenShiftDefaultGatewayClassName, operatorcontroller.OpenShiftGatewayClassControllerName)
+	if err != nil {
+		t.Fatalf("failed to create gatewayclass: %v", err)
+	}
+
+	// Create a Gateway with the listener hostname "foo.[...]"
+	gateway, err := createGatewayWithListeners(gatewayClass, "test-nohost-gateway", operatorcontroller.DefaultOperandNamespace, []struct {
+		Name     string
+		Hostname *string
+	}{
+		{Name: "http-listener-no-host", Hostname: nil},
+	})
+
+	t.Cleanup(func() {
+		gateway := gatewayapiv1.Gateway{ObjectMeta: metav1.ObjectMeta{Name: "test-nohost-gateway", Namespace: operatorcontroller.DefaultOperandNamespace}}
+		if err := kclient.Delete(context.TODO(), &gateway); err != nil {
+			if errors.IsNotFound(err) {
+				return
+			}
+			t.Errorf("failed to delete gateway %q: %v", gateway.Name, err)
+		}
+	})
+
+	if err != nil {
+		t.Fatalf("failed to create gateway with a listener with no hostname: %v", err)
+	}
+
+	gateway, err = assertGatewaySuccessful(t, operatorcontroller.DefaultOperandNamespace, "test-nohost-gateway")
+	if err != nil {
+		t.Fatalf("failed to create the test gateway: %v", err)
+	}
+
+	t.Logf("Created gateway %s with a listener with no hostname", gateway.Name)
+
+	err = wait.PollUntilContextTimeout(context.Background(), 1*time.Second, 1*time.Minute, false, func(context context.Context) (bool, error) {
+		dnsRecords := &iov1.DNSRecordList{}
+		// List all DNSRecords from the default operand namespace.
+		if err := kclient.List(context, dnsRecords, client.InNamespace(operatorcontroller.DefaultOperandNamespace)); err != nil {
+			t.Logf("failed to list DNSRecords: %v; retrying...", err)
+			return false, nil
+		}
+
+		for _, record := range dnsRecords.Items {
+			if record.Labels["gateway.networking.k8s.io/gateway-name"] == gateway.Name {
+				t.Fatalf("dnsrecord found while not expected: %v", err)
+			}
+		}
+
+		t.Logf("No DNSRecord for gateway listener as expected...")
```

Comment:
This is getting all the dnsRecords and trying to find the unexpected one.  It could possibly take a little while to delete it, so it would be better for you to just look for one dnsRecord and poll until it disappears, then fail if it isn't gone in 1 minute.

```suggestion
               dnsRecord := &v1.DNSRecord{}
               recordName := operatorcontroller.GatewayDNSRecordName(gateway, domain)
               
		// Try to get the dnsRecord, and if it's not found, exit the poll successfully.
		if err := kclient.Get(context, recordName, dnsRecord); err != nil {
                        if errors.IsNotFound(err) {
				return true, nil
			}
			// If it is found, then keep trying to see if it gets deleted.
			t.Logf("Unexpected failure getting dnsRecord, retrying....")
		}
	        return false, nil
		})
		
		if err != nil {
		    t.Logf("Expected dnsRecord to be deleted but it wasn't")
                }
		t.Logf("No dnsRecord for gateway listener as expected...")
```

Reply:
In this case our intent was to look if any DNSRecord gets wrongly created for a gateway which has only one listener, which does not have a hostname field.
I think that `GatewayDNSRecordName` would not be appropriate here as we would need to know the DNSRecord name. Also conciously avoided using this function throughout the tests as it is used by the controller and wanted to avoid usage of the tested code. But maybe that was being overly cautious.

Or did I misunderstand and for this test case a valid script would be to:
 - create a gateway with a listener with `foo.domain.` hostname
 - delete the listener/replace it with one with no hostname
 - assert the DNSRecord for `foo.domain.` was deleted
 ?